### ISSUE TITLE
Question about accessing Custom Status from a Durable Orchestration Context

### ISSUE THREAD

**Issue by yukscck** [Azure/azure-functions-durable-extension#947]
Hello,

I am currently trying out the **SetCustomStatus** feature in Durable Functions.

From what I understand, by calling the API, the value we provide to it is saved in CustomStatus of the orchestration instance. Subsequent calls to the API will overwrite the value of that field to the latest one.

Is there any way we can access the current value of the Custom Status from a Durable Orchestration Context, mid execution?

Thanks!

**Comment by olitomlinson** [comment]
@yukscck I don't believe you can access the `CustomStatus` from within an Orchestration Context.

The `customStatus` was designed for serializing transient/ephemeral state within the Orchestrator that external callers might wish to take a read-only dependency on.

May I ask what your use-case is for wanting to access the `customStatus` from within the Orchestration code?

**Comment by yukscck** [comment]
Hi @olitomlinson, thanks for your response.

I'm using Durable Functions for executing a process that is heavy on control flow - depending on the outputs of an Activity (or a few Activities), it would be result to different subsequent execution paths. You can imagine it as a state machine, the execution could jump back and forth between states, until an exit criteria is met.

Due to the dynamic nature of the process, I need to keep track of the execution status - at which point of the process the Orchestration is currently executing.

Looking at the set of APIs Durable Framework offers, I would imagine that `SetCustomStatus` is probably the way to solve this. An HTTP Trigger is created for users to check for the current status of an Orchestration instance, by calling the `GetStatusAsync`.

Here is where I came to the point where I had this question mentioned in this thread.
In my use case, while setting the custom status in an Orchestration Context, the process would first need to know the current status, perform some decision-making, and then update the custom status.

To do so, I explored the option calling the `GetStatusAsync` from the context, which would result to the "multi-threading in Orchestration" issue. But I would imagine delegating this to an Activity function wouldn't work well, because the Orchestration would need to do "Get Status Activity" at every point before executing an Activity that does the actual process-related operations.

Hence, I was curious if I could solve the problem by reading the `CustomStatus` from an Orchestration Context.

Appreciate if you could point out if there is anything I missed.
Thanks!

**Comment by olitomlinson** [comment]
@yukscck Thanks for the details. 

What is stopping you from having a variable in your orchestration that tracks the current state/status, and refer to that instead? My assumption is that you would have this variable anyway as this is what you would use to seed your `customStatus`  when using `setCustomStatus`? 

I do something similar in one of my orchestrations. I have an audit log (which is just essentially a List<Audit>) when significant things happens I append to that audit log, and also refer to the audit when I need to make decisions in the orchestration.

Sorry if I'm not being helpful here, I don't think I fully understand your problem. 

If you can offer some snippets/pseudo of what your code would look like (if such a `getCustomStatus` method were to exist), that might help me understand your scenario a little clearer.

**Comment by yukscck** [comment]
@olitomlinson I agree that passing it as a variable in the orchestration works perfectly if the Orchestration is the only one interested in the status, but I think doing so doesn't cater for the need for users to find out the current status (`GetStatusAsync` via HTTP Trigger).

I could make Orchestration update the variable as you suggested, and then call `SetCustomStatus` as a two step workaround, but it would be great if we could use `CustomStatus` as the source of truth.

**Comment by olitomlinson** [comment]
> I agree that passing it as a variable in the orchestration works perfectly if the Orchestration is the only one interested in the status

Am I correct in understanding that in your use-case you want to use the `customStatus` **outside** of the Durable Function App? 

For example, code executing in a completely different Function App wants to see the `customStatus` of your orchestration? This is already possible using the [Durable Functions HTTP API ](https://docs.microsoft.com/en-us/azure/azure-functions/durable/durable-functions-http-api#get-instance-status)

**Comment by yukscck** [comment]
Hi again, so sorry for not replying, I was caught up with an unrelated urgent issue.

@olitomlinson on your question: yes, I have another service **outside** of Durable Function App, and I have that part covered by exposing an HTTP Trigger in the Function App, which calls `DurableOrchestrationClient.GetStatusAsync`.

The challenge in my scenario is on the OrchestrationContext itself. In the context, the process is required to get to know its current CustomStatus, and then update the CustomStatus.

Understand you have suggested to create a variable in the orchestration to keep track of the value i intended to store in CustomStatus, which would work.
More to my curiousity, I noticed in the `DurableOrchestrationContext` class, there's a private field `serializedCustomStatus`, which already holds the information I need. Just wondering if there's any reason it is not exposed as a public property.

By the way, I noticed this thread has been closed by the bot, do I need to reopen it?

Thanks!
